server/src/pages/shot/editor.js

```diff
@@ -34,23 +51,278 @@ exports.Editor = class Editor extends React.Component {
+      this.cropper.removeEventListener("mousemove", this.mousemove);
+      this.cropper.removeEventListener("mouseout", this.mouseout);
+      this.cropper.removeEventListener("mousedown", this.mousedown);
+      this.cropper.removeEventListener("mouseup", this.mouseup);
+    }
+    this.edit();
+  }
+
+  onClickCrop() {
+    this.setState({tool: 'crop'});
+    this.cropToolBar = <div className="annotation-tools">
+      <button className={`button transparent confirm-crop`} id="confirm-crop" onClick={this.onClickConfirmCrop.bind(this)} title="confirm-crop">Confirm</button>
+      <button className={`button transparent cancel-crop`} id="cancel-crop" onClick={this.onClickCancelCrop.bind(this)} title="cancel-crop">Cancel</button>
+    </div>
+  }
+
+  onClickConfirmCrop() {
+    let x1 = boxes[0].x1;
+    let x2 = boxes[0].x2;
+    let y1 = boxes[0].y1;
+    let y2 = boxes[0].y2;
+    x1 = x1 < 0 ? 0 : x1;
+    y1 = y1 < 0 ? 0 : y1;
+    x2 = x2 > this.canvasWidth ? this.canvasWidth : x2;
+    y2 = y2 > this.canvasHeight ? this.canvasHeight : y2;
+    let cropWidth = x2 - x1;
+    let cropHeight = y2 - y1;
+    let croppedImage = document.createElement('canvas');
+    croppedImage.width = parseInt(cropWidth, 10);
+    croppedImage.height = parseInt(cropHeight, 10);
+    let croppedContext = croppedImage.getContext("2d");
+    croppedContext.drawImage(this.imageCanvas, x1, y1, croppedImage.width, croppedImage.height, 0, 0, croppedImage.width, croppedImage.height);
+    croppedContext.drawImage(this.editor, x1, y1, croppedImage.width, croppedImage.height, 0, 0, croppedImage.width, croppedImage.height);
+    croppedContext.globalCompositeOperation = 'multiply';
+    croppedContext.drawImage(this.highlighter, x1, y1, croppedImage.width, croppedImage.height, 0, 0, croppedImage.width, croppedImage.height);
+    this.canvasWidth = cropWidth;
+    this.canvasHeight = cropHeight;
+    let img = new Image();
+    let imageContext = this.imageCanvas.getContext('2d');
+    img.crossOrigin = 'use-credentials';
+    let width = cropWidth;
+    let height = cropHeight;
+    img.onload = () => {
+      imageContext.drawImage(img, 0, 0, width, height);
+    }
+    this.imageContext = imageContext;
+    img.src = croppedImage.toDataURL("image/png");
+    this.setState({image: 'cropped'});
     this.edit();
+    this.onClickCancelCrop();
   }
 
+  onClickCancelCrop() {
+    this.cropToolBar = null;
+    this.cropper.getContext('2d').clearRect(0, 0, this.cropper.width, this.cropper.height);
+    boxes = [];
+    this.setState({tool: 'pen'});
+  }
+
+  mouseup(e) {
+    if (this.clickedArea.box == -1 && this.tempBox != null && boxes.length < 1) {
+      boxes.push(this.tempBox);
+    } else if (this.clickedArea.box != -1) {
+      let selectedBox = boxes[this.clickedArea.box];
+      if (selectedBox.x1 > selectedBox.x2) {
+        let previousX1 = selectedBox.x1;
+        selectedBox.x1 = selectedBox.x2;
+        selectedBox.x2 = previousX1;
```

Comment:
You could simply do `[selectedBox.x1, selectedBox.x2] = [selectedBox.x2, selectedBox.x1]`;